File: tests/eval/judge.py
"""AI 诊断 LLM-as-judge 评分（T9 + docs/ai-quality.md）。

Per docs/ai-quality.md scoring rubric:
- 5-dim score card (1-5 each): 根因准确性, 修复建议可执行性, 证据链完整性, 幻觉率, 响应时间
- Total = 25, pass threshold = 18 (72%)
- CI gate: 平均 ≥ 18 → pass, 17-18 → warn, < 17 → fail

This script:
1. Loads baseline_samples.json
2. For each sample, runs the AI Agent (mocked in unit tests, real in CI)
3. Sends the AI output to Claude-as-judge for 5-dim scoring
4. Asserts the pass threshold

Run: `pytest -m eval` (CI workflow calls this with ANTHROPIC_API_KEY set).
"""

from __future__ import annotations

import asyncio
import json
import time
from pathlib import Path
from typing import Any

import pytest


PASS_THRESHOLD = 18  # of 25
JUDGE_MODEL = "claude-sonnet-4-5"

JUDGE_SYSTEM_PROMPT = """You are an expert SRE evaluator. You will receive an \
Alibaba Cloud alert and an AI-generated diagnosis. Score the diagnosis on \
5 dimensions (1-5 each):

1. 根因准确性 (root cause accuracy): does it correctly identify the cause?
2. 修复建议可执行性 (recommendation executability): are actions concrete + actionable?
3. 证据链完整性 (evidence chain completeness): does every claim trace to a tool call?
4. 幻觉率 (hallucination rate): 5=no invented IDs/timestamps; 1=many hallucinations
5. 响应时间 (response time): 5=<5s; 1=>30s

Output JSON only:
{"root_cause": 1-5, "recommendation": 1-5, "evidence": 1-5, "hallucination": 1-5, "latency": 1-5, "reasoning": "..."}
"""


def load_baseline_samples() -> list[dict[str, Any]]:
    """Load the 10 baseline samples from JSON."""
    path = Path(__file__).parent / "baseline_samples.json"
    return json.loads(path.read_text())["samples"]


@pytest.fixture(scope="module")
def baseline_samples() -> list[dict[str, Any]]:
    return load_baseline_samples()


@pytest.mark.eval
def test_baseline_samples_exist() -> None:
    """Sanity check: 10 samples loaded."""
    samples = load_baseline_samples()
    assert len(samples) == 10, f"expected 10 baseline samples, got {len(samples)}"


@pytest.mark.eval
@pytest.mark.asyncio
async def test_each_sample_meets_threshold(baseline_samples: list[dict[str, Any]]) -> None:
    """For each sample, run AI Agent + LLM-judge. Assert mean score ≥ 18.

    This test is the CI gate. It requires ANTHROPIC_API_KEY.
    """
    # M1: this test imports the agent and runs end-to-end. Skipped in unit tests.
    pytest.skip("M1: requires AI Agent implementation (T2) before this can run end-to-end")


def _stub_score_for_unit_tests() -> dict[str, int]:
    """A passing stub score for unit-testing the judge logic itself."""
    return {
        "root_cause": 4,
        "recommendation": 4,
        "evidence": 3,
        "hallucination": 5,
        "latency": 4,
        "reasoning": "stub for unit tests",
    }


def aggregate_scores(scores: list[dict[str, int]]) -> dict[str, float]:
    """Compute mean across all 5 dimensions for a list of judge outputs."""
    if not scores:
        return {k: 0.0 for k in ("root_cause", "recommendation", "evidence", "hallucination", "latency")}
    return {
        k: sum(s[k] for s in scores) / len(scores)
        for k in ("root_cause", "recommendation", "evidence", "hallucination", "latency")
    }


def total_score(s: dict[str, int]) -> int:
    """Sum the 5-dim score card (max 25)."""
    return sum(s[k] for k in ("root_cause", "recommendation", "evidence", "hallucination", "latency"))


def test_judge_logic_passes_stub() -> None:
    """Sanity: the stub score passes the threshold."""
    assert total_score(_stub_score_for_unit_tests()) >= PASS_THRESHOLD


def test_aggregate_scores_empty() -> None:
    """Empty input → zeros, no division error."""
    agg = aggregate_scores([])
    assert agg["root_cause"] == 0.0
    assert agg["recommendation"] == 0.0


@pytest.mark.eval
def test_ci_gate_evaluation(baseline_samples: list[dict[str, Any]]) -> None:
    """The actual CI gate. Will run when ANTHROPIC_API_KEY is set + AI Agent ready.

    Logic placeholder: aggregate scores and fail CI if mean total < PASS_THRESHOLD.
    """
    pytest.skip("M1: gate implementation deferred to AI Agent integration (T2)")

    # Future:
    # scores = []
    # for sample in baseline_samples:
    #     diagnosis = await ai_agent.diagnose(sample["alert_payload"])
    #     judge_output = await judge_model.score(sample, diagnosis)
    #     scores.append(judge_output)
    # mean_total = sum(total_score(s) for s in scores) / len(scores)
    # assert mean_total >= PASS_THRESHOLD, f"mean={mean_total:.1f} < {PASS_THRESHOLD}"

File: internal/eval/judge.go
package eval

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"

	"github.com/anthropics/anthropic-sdk-go"
	"github.com/anthropics/anthropic-sdk-go/option"

	"github.com/buhaiqing/ai-cloud-ops/internal/agent"
)

// JudgeSystemPrompt mirrors the Python reference. JSON-only output with the
// five score keys plus a free-form reasoning field.
const JudgeSystemPrompt = `You are an expert SRE evaluator. You will receive an ` +
	`Alibaba Cloud alert and an AI-generated diagnosis. Score the diagnosis on ` +
	`5 dimensions (1-5 each):

1. 根因准确性 (root cause accuracy): does it correctly identify the cause?
2. 修复建议可执行性 (recommendation executability): are actions concrete + actionable?
3. 证据链完整性 (evidence chain completeness): does every claim trace to a tool call?
4. 幻觉率 (hallucination rate): 5=no invented IDs/timestamps; 1=many hallucinations
5. 响应时间 (response time): 5=<5s; 1=>30s

Output JSON only:
{"root_cause": 1-5, "recommendation": 1-5, "evidence": 1-5, "hallucination": 1-5, "latency": 1-5, "reasoning": "..."}
`

// anthropicClient is the subset of the Anthropic SDK used by Judge. Defined
// here so tests can inject a fake without a live network.
type anthropicClient interface {
	CreateMessage(ctx context.Context, body anthropic.MessageNewParams) (*anthropic.Message, error)
}

// realAnthropicClient wraps anthropic.Client (a value, not a pointer in v1.x)
// to satisfy the anthropicClient interface.
type realAnthropicClient struct{ c anthropic.Client }

// CreateMessage delegates to the SDK.
func (r *realAnthropicClient) CreateMessage(ctx context.Context, body anthropic.MessageNewParams) (*anthropic.Message, error) {
	return r.c.Messages.New(ctx, body)
}

// Judge asks Claude to score a diagnosis on the 5-dim rubric.
type Judge struct {
	anthropicClient anthropicClient
	model           string
	log             *slog.Logger
}

// NewJudge builds a Judge backed by the real Anthropic API.
func NewJudge(apiKey string) *Judge {
	c := anthropic.NewClient(option.WithAPIKey(apiKey))
	return newJudgeWith(&realAnthropicClient{c: c}, slog.Default())
}

// newJudgeWith is the internal constructor; tests use it to inject fakes.
func newJudgeWith(c anthropicClient, log *slog.Logger) *Judge {
	if log == nil {
		log = slog.Default()
	}
	return &Judge{anthropicClient: c, model: JudgeModel, log: log}
}

// rawScore is the JSON wire shape. Strict parsing (DisallowUnknownFields) makes
// upstream drift visible: a new dimension on the model side → explicit error.
type rawScore struct {
	RootCause      int    `json:"root_cause"`
	Recommendation int    `json:"recommendation"`
	Evidence       int    `json:"evidence"`
	Hallucination  int    `json:"hallucination"`
	Latency        int    `json:"latency"`
	Reasoning      string `json:"reasoning"`
}

func (r rawScore) toScoreCard() ScoreCard {
	return ScoreCard{
		DimRootCause:      r.RootCause,
		DimRecommendation: r.Recommendation,
		DimEvidence:       r.Evidence,
		DimHallucination:  r.Hallucination,
		DimLatency:        r.Latency,
	}
}

// Score queries the judge model and returns the parsed ScoreCard.
func (j *Judge) Score(ctx context.Context, alert map[string]any, diagnosis *agent.Diagnosis) (ScoreCard, error) {
	body := anthropic.MessageNewParams{
		Model:     j.model,
		MaxTokens: 1024,
		System: []anthropic.TextBlockParam{
			{Text: JudgeSystemPrompt},
		},
		Messages: []anthropic.MessageParam{
			anthropic.NewUserMessage(anthropic.NewTextBlock(buildUserPrompt(alert, diagnosis))),
		},
	}
	resp, err := j.anthropicClient.CreateMessage(ctx, body)
	if err != nil {
		return nil, fmt.Errorf("judge: create message: %w", err)
	}
	raw := extractText(resp)
	if raw == "" {
		return nil, fmt.Errorf("judge: empty model response")
	}
	raw = stripCodeFences(raw)
	var parsed rawScore
	dec := json.NewDecoder(strings.NewReader(raw))
	dec.DisallowUnknownFields()
	if err := dec.Decode(&parsed); err != nil {
		return nil, fmt.Errorf("judge: parse score json: %w (raw=%s)", err, raw)
	}
	card := parsed.toScoreCard()
	for _, d := range Dims() {
		v := card[d]
		if v < 1 || v > 5 {
			return nil, fmt.Errorf("judge: dim %s out of range 1-5: got %d", d, v)
		}
	}
	return card, nil
}

func buildUserPrompt(alert map[string]any, d *agent.Diagnosis) string {
	alertJSON, err := json.MarshalIndent(alert, "", "  ")
	if err != nil {
		alertJSON = fmt.Appendf(nil, "%v", alert)
	}
	diagJSON, err := json.MarshalIndent(d, "", "  ")
	if err != nil {
		diagJSON = fmt.Appendf(nil, "%v", d)
	}
	return fmt.Sprintf("Alert:\n%s\n\nAI Diagnosis:\n%s", alertJSON, diagJSON)
}

func extractText(msg *anthropic.Message) string {
	var b strings.Builder
	for _, block := range msg.Content {
		// Prefer explicit text-type; fall back to any non-empty Text payload so
		// downstream parsing is robust against partial/streaming responses.
		if block.Type == "text" || block.Text != "" {
			b.WriteString(block.Text)
		}
	}
	return b.String()
}

// stripCodeFences removes ```json ... ``` fences if present. Model output
// occasionally wraps JSON in a code block; we accept both forms.
func stripCodeFences(s string) string {
	s = strings.TrimSpace(s)
	if !strings.HasPrefix(s, "```") {
		return s
	}
	s = strings.TrimPrefix(s, "```")
	// optional language tag on first line
	if i := strings.IndexByte(s, '\n'); i >= 0 {
		s = s[i+1:]
	}
	if j := strings.LastIndex(s, "```"); j >= 0 {
		s = s[:j]
	}
	return strings.TrimSpace(s)
}
